core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,715 @@
+			caller->job_id = job_mask;
+			caller->jobctx = jobctx;
+			caller->pdesc = virt_to_phys((void *)jobctx->desc);
+
+			found = true;
+		}
+	}
+	cpu_spin_unlock(&jr_privdata->callers_lock);
+
+	if (!found) {
+		JR_TRACE("Error didn't find a free space in the callers array");
+		goto end_enqueue;
+	}
+
+	JR_TRACE("Push id=%d, job (0x%08x) context @0x%08" PRIxVA,
+		 jr_privdata->inwrite_index, job_mask, (vaddr_t)jobctx);
+
+	cur_inrings = &jr_privdata->inrings[jr_privdata->inwrite_index];
+
+	/* Push the descriptor into the JR HW list */
+	caam_write_val(cur_inrings, caller->pdesc);
+
+	/* Ensure that physical memory is up to date */
+	cache_operation(TEE_CACHECLEAN, (void *)cur_inrings,
+			sizeof(struct inring_entry));
+
+	/*
+	 * Increment index to next JR input entry taking care that
+	 * is a circular buffer of nb_jobs size.
+	 */
+	jr_privdata->inwrite_index++;
+	jr_privdata->inwrite_index %= jr_privdata->nb_jobs;
+
+	/* Ensure that input descriptor is pushed in physical memory */
+	cache_operation(TEE_CACHECLEAN, (void *)jobctx->desc,
+			DESC_SZBYTES(caam_desc_get_len(jobctx->desc)));
+
+	/* Inform HW that a new JR is available */
+	caam_hal_jr_add_newjob(jr_privdata->baseaddr);
+
+	*job_id = job_mask;
+	retstatus = CAAM_NO_ERROR;
+
+end_enqueue:
+	cpu_spin_unlock_xrestore(&jr_privdata->inlock, exceptions);
+
+	return retstatus;
+}
+
+/*
+ * Synchronous job completion callback
+ *
+ * @jobctx   Job context
+ */
+static void job_done(struct caam_jobctx *jobctx)
+{
+	jobctx->completion = true;
+}
+
+/*
+ * Cancels a job id. Remove the job from SW Job array
+ *
+ * @job_id     Job Id
+ */
+void caam_jr_cancel(uint32_t job_id)
+{
+	uint8_t idx = 0;
+
+	JR_TRACE("Job cancel 0x%" PRIx32, job_id);
+	for (idx = 0; idx < jr_privdata->nb_jobs; idx++) {
+		/*
+		 * Search for the caller information corresponding to
+		 * the job_ids mask.
+		 */
+		if (jr_privdata->callers[idx].job_id == job_id) {
+			/* Clear the Entry Descriptor */
+			jr_privdata->callers[idx].pdesc  = 0;
+			jr_privdata->callers[idx].job_id = JR_JOB_FREE;
+			return;
+		}
+	}
+}
+
+/*
+ * Checks if one of the given job ids in bit mask format
+ * is complete. If none are complete, wait until timeout expired.
+ * Wait infinitively is @timeout_ms = (-1)
+ *
+ * @job_ids     Job Ids Mask
+ * @timeout_ms  Timeout in millisecond
+ */
+enum CAAM_Status caam_jr_dequeue(uint32_t job_ids, uint32_t timeout_ms)
+{
+	uint32_t jobComplete = 0;
+	uint32_t nbTimeWait = 0;
+	bool infinite = false;
+
+	if (timeout_ms == UINT32_MAX) {
+		infinite = true;
+	} else {
+		/* Divide the timeout_ms in 10 ms wait time */
+		nbTimeWait = timeout_ms / 10;
+	}
+
+	do {
+		/* Call the do_jr_dequeue function to dequeue the jobs */
+		jobComplete = do_jr_dequeue(job_ids);
+
+		if (jobComplete & job_ids)
+			return CAAM_NO_ERROR;
+
+		/* Check if JR interrupt otherwise wait a bit */
+		if (!caam_hal_jr_poolack_itr(jr_privdata->baseaddr))
+			caam_udelay(10);
+	} while (infinite || (nbTimeWait--));
```

Comment:
In case timeout_ms = 0, we execute at least the job. I don't want to add a check if (timeout_ms == 0) ....
There is Real Time precision in the timeout. The value is used to not be locked in infinite loop when not necessary and this is a maximum waiting time before saying there is a problem.

Reply:
That's a problem with the do - while pattern, it often leads to convoluted expressions. As long as it doesn't spread into the rest of the OP-TEE code base I guess I can turn a blind a eye to it here.